internal/authentication/ldap_client_factory.go

```diff
@@ -117,162 +121,393 @@ type PooledLDAPClientFactory struct {
+		return f.disposeClient(pooled)
 	}
 
-	return nil
+	pC := f.pool
+	if pC == nil {
+		logging.Logger().Warning("LDAP pool is not initialized or was closed, disposing client")
+		return f.disposeClient(pooled)
+	}
+
+	// NOTE: clients can still be returned while the pool is closing.
+	// Rationale: cleanup in poolManager() handles this gracefully.
+
+	select {
+	case pC <- pooled:
+		logging.Logger().Trace("Successfully returned client to LDAP pool")
+		return nil
+	default: // shouldn't happen, unless returning more clients than acquired
+		logging.Logger().Debug("LDAP pool is full, disposing client")
+		return f.disposeClient(pooled)
+	}
 }
 
-func (f *PooledLDAPClientFactory) acquire(ctx context.Context) (client *LDAPClientPooled, err error) {
-	f.mu.Lock()
+// thread-safe, closes the client and decrements active count
+func (f *PooledLDAPClientFactory) disposeClient(client *LDAPClientPooled) error {
+	if client == nil {
+		logging.Logger().Warn("Cannot dispose nil LDAP client")
+		return nil
+	}
 
-	defer f.mu.Unlock()
+	remaining := f.activeCount.Add(-1)
+	logging.Logger().Tracef("Pooled LDAP client disposed; remaining: %d", remaining)
+	return client.Close()
+}
 
-	if f.closing {
-		return nil, fmt.Errorf("error acquiring client: the pool is closed")
+// thread-safe, returns a client using the pool or an error
+func (f *PooledLDAPClientFactory) acquire() (client *LDAPClientPooled, err error) {
+	logging.Logger().Trace("Acquiring client from LDAP pool")
+
+	pC := f.pool
+	if pC == nil || f.isClosing() {
+		logging.Logger().Error("Cannot acquire client: LDAP pool is not initialized or closing")
+		return nil, fmt.Errorf("error acquiring client: LDAP pool is not initialized or closing")
 	}
 
-	if cap(f.pool) != f.config.Pooling.Count {
-		if err = f.Initialize(); err != nil {
-			return nil, err
+	now := time.Now()
+	deadline := now.Add(f.config.Pooling.Timeout)
+	interval := time.Millisecond * 10
+
+	for time.Now().Before(deadline) {
+		f.wakeupPoolManager() // request more clients if needed
+		select {
+		case <-time.After(interval): // this avoids the blocking time.Sleep()
+			continue // do nothing, except to loop again to signal the pool manager
+		case client = <-pC:
+			if !client.IsHealthy() { // NOTE: we dispose expired clients only on release
+				logging.Logger().Info("Received invalid client from LDAP pool, disposing")
+				_ = f.disposeClient(client)
+				continue
+			}
+			// NOTE: we do NOT check for expiration here - it's a randomized "soft" target.
+			// Rationale: expiration is ony checked on release to avoid potential starvation.
+			// Clients may be used slightly beyond their expiration time, but this is acceptable.
+			logging.Logger().Tracef("Successfully acquired valid client from LDAP pool after %s", time.Since(now))
+			return client, nil
 		}
 	}
+	return nil, fmt.Errorf("timeout waiting for client from LDAP pool")
+}
+
+// must be called EXCLUSIVELY by Initialize() or poolManager()
```

Comment:
Must begin with a capital and end in a period, also there is no need to put other emphasis or punctuation in comments, please make `EXCLUSIVELY` become `exclusively`.  It may also be beneficial to name the function after the function that should be calling it.